chains/ethereum/contract/ETH/LockRedeem_Kratos.sol

```diff
@@ -0,0 +1,271 @@
+
+        }
+        // Trasaction Cost is a an average trasaction cost .
+        // additionalGasCost is the extra gas used by the lines of code after gas calculation is done.
+        // This is an approximate calcualtion and actuall cost might vary sightly .
+
+        uint gasUsed = startGas - gasleft() + transactionCost + additionalGasCost;
+        uint gasFee = gasUsed * tx.gasprice + (rewardGas * validatorEarningMultiplier);
+        (bool success,) = msg.sender.call.value(gasFee)("");
+        require(success, "Transfer back to validator failed");
+        redeemRequests[recipient_].redeemFee -= gasFee;
+        emit ValidatorSignedRedeem(recipient_, msg.sender, amount_, gasFee);
+    }
+
+
+    function collectUserFee() public isActive {
+        require(verifyRedeem(msg.sender) > 0, "request signing is still in progress");
+        (bool success,) = msg.sender.call.value(redeemRequests[msg.sender].redeemFee)("");
+        require(success, "Transfer failed.");
+    }
+
+    function isValidator(address addr) public view returns (bool) {
+        return validators[addr] > 0;
+    }
+
+    function isredeemAvailable(address recepient_) public isActive view returns (bool)  {
+        return redeemRequests[recepient_].until < block.number;
+    }
+
+    function hasValidatorSigned(address recipient_) public isActive view returns (bool) {
+        return ((redeemRequests[recipient_].votes >> validators[msg.sender]) % 2 == uint256(1));
+    }
+
+    //BeforeRedeem : -1  (amount == 0 and until = 0)
+    //Ongoing : 0  (amount > 0 and until > block.number)
+    //Success : 1  (amount = 0 and until < block.number)
+    //Expired : 2  (amount > 0 and until < block.number)
+    function verifyRedeem(address recipient_) public isActive view returns (int8){
+        return (redeemRequests[recipient_].until > 0 && redeemRequests[recipient_].until < block.number) ? (redeemRequests[recipient_].amount > 0 ? int8(2) : int8(1)) : (redeemRequests[recipient_].amount == 0 ? int8(- 1) : int8(0));
+    }
+
+    function getSignatureCount(address recipient_) public isActive view returns (uint256){
+        return redeemRequests[recipient_].signature_count;
+    }
+
+    function getRedeemBalance(address recipient_) public isActive view returns (uint256){
+        return redeemRequests[recipient_].redeemFee;
+    }
+
+    function getTotalEthBalance() public view returns (uint) {
+        return address(this).balance;
+    }
+
+    function getOLTEthAddress() public view returns (address){
+        return address(this);
+    }
+
+    // internal functions
+    function addValidator(address v) internal {
+        validatorList.push(v);
+        validators[v] = uint8(validatorList.length);
+        emit AddValidator(v);
+    }
+
+    function MigrateFromOld() public {
+        require(msg.sender == old_contract);
+        migrationsignaturecount = migrationsignaturecount + 1;
+        //addValidator(tx.origin);
```

Comment:
why this is removed?

Reply:
after removing it, how can validator be added to the new contract?

Reply:
We are setting up new validators for kratos , so the new Validators will be added throiugh the constructor